src/modules/Unity/Application/mirsurface.cpp

```diff
@@ -537,8 +539,9 @@ void MirSurface::resize(int width, int height)
 
     bool mirSizeIsDifferent = width != m_size.width() || height != m_size.height();
 
-    if (mirSizeIsDifferent) {
+    if (mirSizeIsDifferent || m_sizePendingChange) {
```

Comment:
Nice catch!

The plan with `mirSizeIsDifferent` was to avoid calling resize multiple times with the same value, since QML can call it multiple times per frame when its scene is settling down, creating unnecessary resize events.

But you're right, it is faulty when dropping frames, as a sequence like MirSurface::resize(A,B), MirSurface::resize(B,A), MirSurface::dropPendingBuffer(A,B), MirSurface::resize(A,B) would result in the
last resize being ignored.

The boolean should indeed work. 

I have a niggling worry about it being accessed in two separate threads, updateTexture is called by MirSurfaceItem::updatePaintNode which is on the renderer thread, but resize will be called on the GUI thread. You can see there's a mutex there, but we're not guarding `m_size`, so maybe it's ok. Please test thoroughly :)


I might suggest adding a class member `QSize m_lastResize" and have MirSurface::resize do
```
QSize newSize(width, height);
bool sizeChanged = m_lastResize == newSize;

if (sizeChanged) {
    m_controller->resize(m_window, newSize;
    m_lastResize = newSize;
}
```
which should remove need for `m_sizePendingChange` everywhere else. But I can't test that easily, and stopping comparison with the actual size might have an unexpected consequence.

Overall, a +1 from me

Reply:
I did consider a method like that, but what i came to is try to avoid assuming mir's sizes. I was thinking since mir might resize outside of qtmir where qtmir would still believe its old sizes? but im not sure about the inner details here.

Reply:
That way does indeed lie madness. We can't have multiple possibly conflicting decision points. QtMir implements a Mir WindowManager, and so should be aware of any decisions by Mir and be able to override them. So consider Mir & QtMir working closely together, and collaborating on all decisions, and if there's any disagreement between Mir & QtMir's window models, it's a bug.

Anyway, what you landed should do the trick